src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,244 @@
+                    out string? cancellationTokenParameterName,
+                    out string? methodName))
+                {
+                    return;
+                }
+
+                context.ReportDiagnostic(context.Operation.CreateDiagnostic(ForwardCancellationTokenToAsyncMethodsRule, cancellationTokenParameterName!, methodName!));
+            },
+            OperationKind.Invocation);
+        }
+
+        private static bool ShouldAnalyze(
+            IInvocationOperation invocation,
+            INamedTypeSymbol cancellationTokenType,
+            out string? cancellationTokenParameterName,
+            out string? methodName)
+        {
+            IMethodSymbol method = invocation.TargetMethod;
+
+            cancellationTokenParameterName = null;
+            methodName = method.Name;
+
+            // Check if the invocation has an optional implicit ct or an overload that takes one ct
+            if (!InvocationIgnoresOptionalCancellationToken(cancellationTokenType, invocation, method) &&
+                !MethodHasCancellationTokenOverload(cancellationTokenType, method))
+            {
+                return false;
+            }
+
+            // Find the ancestor method that contains this invocation
+            if (!TryGetAncestorDeclaration(invocation, out IMethodSymbol? methodDeclaration))
+            {
+                return false;
+            }
+
+            // Check if the ancestor method has a ct that we can pass to the invocation
+            if (!VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(cancellationTokenType, methodDeclaration!, out cancellationTokenParameterName))
+            {
+                return false;
+            }
+
+            return cancellationTokenParameterName != null && methodName != null;
+        }
+
+        // Looks for an ancestor that could be a method or function declaration.
+        private static bool TryGetAncestorDeclaration(IInvocationOperation invocation, out IMethodSymbol? declaration)
+        {
+            declaration = null;
+
+            IOperation currentOperation = invocation.Parent;
+            while (currentOperation != null)
+            {
+                if (currentOperation.Kind == OperationKind.AnonymousFunction && currentOperation is IAnonymousFunctionOperation anonymousFunction)
+                {
+                    declaration = anonymousFunction.Symbol;
+                    break;
+                }
+                else if (currentOperation.Kind == OperationKind.LocalFunction && currentOperation is ILocalFunctionOperation localFunction)
+                {
+                    declaration = localFunction.Symbol;
+                    break;
+                }
+                else if (currentOperation.Kind == OperationKind.MethodBody && currentOperation is IMethodBodyOperation methodBody)
+                {
+                    if (methodBody.SemanticModel.GetDeclaredSymbol(methodBody.Syntax) is IMethodSymbol method)
```

Comment:
Please do not use Semantic model APIs when working in IOperation space. All symbolic information should be available in the operation tree. For example, the `OperationAnalysisContext` exposes [ContainingSymbol](http://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,1383), which is the symbol you are computing here. Instead of invoking `GetDeclaredSymbol`, you can pass in `context.ContainingSymbol` as the outermost method containing symbol.

Reply:
Thanks for letting me know. I'll get this updated.

Reply:
I assume I should change the `context.RegisterCompilationStartAction` in the `Initialize` method to now be `context.RegisterOperationStartAction`, if I want to access the `OperationAnalysisContext`. Is this correct or not?

Reply:
This still needs to be address so we avoid calling `GetDeclaredSymbol` APIs in the analyzer.

Reply:
@mavasani the root cause of one of the edge cases is the use of `ContainingSymbol`. Take this code as an example:

```cs
using System;
using System.Threading;
public static class Extensions
{
    public static void Extension(this bool b, Action<int> action) {}
    public static void MyMethod(this int i, CancellationToken c = default) {}
}
class C
{
    public void M(CancellationToken ct)
    {
        bool b = false;
        b.Extension((j) =>
        {
            Console.WriteLine(""Hello world"");
            j.[|MyMethod|]();
        });
    }
}
```

- The analyzer will detect `public void M(CancellationToken ct)` as the containing symbol.
- When the fixer checks the parent, it's `b.Extension(j)` that gets detected as the containing symbol (because I make sure to detect lambdas).
- The fixer assumes the analyzer did its job of checking if the containing symbol had a CancellationToken parameter, and it did - but it was `M`, not the lambda.
- I realized I don't have an easy way to detect the type of `j`. So the fixer currently assumes that `j` is the CancellationToken parameter, and that's what gets passed to `j.MyMethod()` -> `j.MyMethod(j)`, which is clearly wrong.

Do I have an alternative to using `ContainingSymbol`? Is it really bad if I use the semantic model in the analyzer?

Reply:
Manish explained this to me via chat and I addressed it.